### PR TITLE
Try adding additional auth fields

### PR DESCRIPTION
**Pull request by bloodyowl**

### CODE REVIEW

File: src/grape/hooks/auth_field.clj [hunk 1]
```diff
@@ -7,60 +7,40 @@
       (let [auth-strategy (:auth-strategy resource)
             auth-field? (= (:type auth-strategy) :field)
             doc-field (:doc-field auth-strategy)
+            additional-permission-doc-fields (or (:additional-permission-doc-fields auth-strategy) [])
```
Comment (bloodyowl) [inline src/grape/hooks/auth_field.clj]: The idea is to provide a list of additional keys for which the auth is able to authorise edition.

File: src/grape/hooks/auth_field.clj [hunk 2]
```diff
@@ -7,60 +7,40 @@
...
-                                     (if-not existing-value
-                                       auth-value
-                                       existing-value)))))
-   :pre-update-post-validate
-   (wrap-auth :update (fn [auth-value doc-field payload]
-                        (update-in payload [doc-field]
-                                   (fn [existing-value]
-                                     (if (and existing-value (not= (str existing-value) (str auth-value)))
-                                       (throw (ex-info "Forbidden" {:type :forbidden}))
-                                       auth-value)))))
```
Comment (bloodyowl) [inline src/grape/hooks/auth_field.clj]: Reparenting resources seems like a possible bug, so I removed that

File: src/grape/hooks/auth_field.clj [hunk 3]
```diff
@@ -7,60 +7,40 @@
...
+   (wrap-auth :update (fn [auth-value doc-field additional-permission-doc-fields payload]
+                        (if (has-rights-over payload doc-field additional-permission-doc-fields auth-value)
+                              payload
+                              (throw (ex-info "Forbidden" {:type :forbidden})))))
    :pre-read
-   (wrap-auth :read (fn [auth-value doc-field query]
+   (wrap-auth :read (fn [auth-value doc-field additional-permission-doc-fields query]
                       (update-in query [:find]
                                  (fn [find]
                                    (let [existing-value (doc-field find)
```
Comment (bloodyowl) [inline src/grape/hooks/auth_field.clj]: The pre-read seems a little more tricky here, not sure what to do

File: src/grape/hooks/auth_field.clj (hunk 2 above)
Comment (cyppan) [reply src/grape/hooks/auth_field.clj]: a bug ? It isn't enabled on partial update

**Comment by cyppan** [comment]
The original idea was to let the user use pre-defined authorization strategies, this one (auth-field) being very simple, but it still allows the connected account id to be automatically injected when not specified in the payload.
For more complex use cases, it's indeed better to skip the injection mechanism, and maybe let the application using grape define something like `:auth {:write-predicate (fn [deps payload] true / false) :read-wrap-query (fn [deps query] authenticatedQuery)}` in the resource spec.
Still, your additional-fields approach work, see my comment below for handling the read part

File: src/grape/hooks/auth_field.clj (hunk 2 above)
Comment (bloodyowl) [reply src/grape/hooks/auth_field.clj]: Oh right, I forgot that `update` hooks only mached `PUT` requests.

**Comment by cyppan** [comment]
And regarding the :read part, it's probably better to wrap the provided mongo find entirely like:
`query = {"$and": [yourAuthScoping, originalQuery]}`
Not sure what yourAuthScoping would be in your case though

**Pull request closed without merging**